eva_submission/eload_ingestion.py

```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
             self.error('ENA metadata load failed: aborting ingestion.')
             self.eload_cfg.set(self.config_section, 'ena_load', value='failure')
             raise e
+
+    def setup_project_dir(self):
+        """
+        Sets up project directory and copies VCF files from the eload directory.
+        """
+        project_dir = Path(cfg['projects_dir'], self.project_accession)
+        os.makedirs(project_dir, exist_ok=True)
+        for _, v in project_dirs.items():
+            os.makedirs(project_dir.joinpath(v), exist_ok=True)
+        # copy valid vcfs + index to 'valid' folder
+        valid_dir = project_dir.joinpath(project_dirs['valid'])
+        vcf_dict = self.eload_cfg.query('brokering', 'vcf_files')
+        if vcf_dict is None:
+            self.error('No brokered VCF files found, aborting ingestion.')
+            raise ValueError('No brokered VCF files found.')
+        for key, val in vcf_dict.items():
+            vcf_path = Path(key)
+            shutil.copyfile(vcf_path, valid_dir.joinpath(vcf_path.name))
+            index_path = Path(val['index'])
+            shutil.copyfile(index_path, valid_dir.joinpath(index_path.name))
+        self.eload_cfg.set(self.config_section, 'project_dir', value=str(project_dir))
+        return project_dir
+
+    def merge_vcfs(self):
+        """
+        Attempts to merge valid VCFs by sample.
+        If duplicate sample names are detected, this will assume files are split horizontally
+        and so no merge is necessary.
+        Returns the list of filenames that should be used for the load pipeline, i.e. either
+        a singleton list with the merged filename, or the list of valid VCFs.
+        """
+        merged_dir = self.project_dir.joinpath(project_dirs['merged'])
+        output_file = str(merged_dir.joinpath(f'{self.project_accession}_merged.vcf'))
+        list_file = str(merged_dir.joinpath('all_files.list'))
+        with open(list_file, 'w+') as f:
+            f.write('\n'.join(str(fn) for fn in self.valid_vcf_filenames))
+        try:
+            command_utils.run_command_with_output(
+                'Merge VCFs',
+                ' '.join((
+                    cfg['executable']['bcftools'], 'merge'
+                    '--merge', 'all',
+                    '--file-list', list_file,
+                    '--threads', '3',
+                    '-o', output_file
+                ))
+            )
+            command_utils.run_command_with_output(
+                'Bgzip merged vcf file',
+                ' '.join((cfg['executable']['bgzip'], output_file))
+            )
+            return [Path(f'{output_file}.gz')]
+        except subprocess.CalledProcessError as e:
+            if e.output is not None and 'Duplicate sample names' in e.output:  # TODO brittle
+                self.warning('Duplicate sample names found while merging, will continue with unmerged VCFs.')
```

Comment:
I am not sure if it is prudent to do this... Let's discuss further in DEV meeting...

Reply:
Captured in [EVA-2334](https://www.ebi.ac.uk/panda/jira/browse/EVA-2334), will update the `TODO` to refer to the ticket